pages/doc.js

```diff
@@ -25,216 +24,146 @@ import { media } from '../src/styles'
-
-    if (!item) {
-      this.setState({ pageNotFound: true, currentItem: {} })
-    } else if (!isFirstPage && !isPageChanged) {
-      this.updateScroll(isPageChanged)
-    } else {
-      this.setState({ isLoading: true, headings: [] })
-      fetch(item.source)
-        .then(res => {
-          res.text().then(text => {
-            this.setState(
-              {
-                currentItem: item,
-                headings: this.parseHeadings(text),
-                isLoading: false,
-                isMenuOpen: false,
-                markdown: text,
-                pageNotFound: false
-              },
-              () => this.updateScroll(!isFirstPage && isPageChanged)
-            )
-          })
-        })
-        .catch(() => {
-          window.location.reload()
-        })
-    }
-  }
+        <Side isOpen={isMenuOpen}>
+          {isSearchAvaible && (
+            <SearchArea>
+              <SearchForm />
+            </SearchArea>
+          )}
 
-  updateScroll(isPageChanged) {
-    const { hash } = window.location
+          <SidebarMenu sidebar={sidebar} currentPath={path} id={SIDEBAR_MENU} />
+        </Side>
+
+        <Markdown
+          markdown={markdown}
+          githubLink={githubLink}
+          prev={prev}
+          next={next}
+          tutorials={tutorials}
+        />
+
+        <RightPanel
+          headings={headings}
+          githubLink={githubLink}
+          tutorials={tutorials}
+        />
+      </Container>
+    </Page>
+  )
+}
 
-    if (isPageChanged) {
-      this.setState({ isSmoothScrollEnabled: false }, () => {
-        this.scrollTop()
-        this.setState({ isSmoothScrollEnabled: true }, () => {
-          if (hash) this.scrollToLink(hash)
-        })
-      })
-    } else if (hash) {
-      this.scrollToLink(hash)
+Documentation.getInitialProps = async ({ asPath, req }) => {
+  const item = getItemByPath(asPath)
+
+  if (!item) {
+    return {
+      errorCode: 404
     }
   }
 
-  parseHeadings = text => {
-    const headingRegex = /\n(## \s*)(.*)/g
-    const matches = []
-    let match
-    do {
-      match = headingRegex.exec(text)
-      if (match)
-        matches.push({
-          text: match[2],
-          slug: kebabCase(match[2])
-        })
-    } while (match)
-
-    return matches
-  }
+  const host = req ? req.headers['host'] : window.location.host
+  const protocol = host.indexOf('localhost') > -1 ? 'http:' : 'https:'
 
-  scrollToLink = hash => {
-    const element = document.getElementById(hash.replace(/^#/, ''))
+  try {
+    const res = await fetch(`${protocol}//${host}${item.source}`)
 
-    if (element) {
-      const headerHeight = document.getElementById(HEADER).offsetHeight
-      const elementBoundary = element.getBoundingClientRect()
-      const rootElement = document.getElementById(ROOT_ELEMENT)
-      rootElement.scroll({ top: elementBoundary.top - headerHeight })
+    if (res.status !== 200) {
+      return {
+        errorCode: res.status
+      }
```

Comment:
I'd just return a 504 here.

Reply:
Maybe with an error message including the `res.status`.

Reply:
Why not leave actual code there? It can help us debug things. Do you see some case there showing real code can be bad?

Reply:
Ah, this fetch is for the internal document? Not some sort of gateway? In that case I think returning actual code makes sense indeed. Although a generic code with a message explaining what happened (and the source error code) would make ever more sense I think. This is merged and works though, so whatevs.